### PR TITLE
Groups backend support

### LINKED ISSUE #31: Add support for groups
### Add support for groups

A common use case for file permissions is to have groups of people with same access to specific files/folders. It would be a great addition to the Datastore if we could support something similar.

### Contract interface proposal

- createGroup(string _group, address[] _entities)
- deleteGroup(string _group)
- renameGroup(string _group, string _newGroupName)
- addEntityToGroup(string _group, address _entity)
- removeEntityFromGroup(string _group, address _entity)
- setGroupPermissions(uint _fileId, string _group, bool _read, bool _write)
- removeGroupFromFile(uint _fileId, string _group)

### PR DESCRIPTION
**Pull request by juslar**
Fix #31

### CODE REVIEW

File: contracts/Datastore.sol [hunk 1]
```diff
@@ -329,6 +341,147 @@ contract Datastore {
...
+        return false;
+    }
+
+    /**
+     * @notice Add a group to the datastore
+     * @param _groupName Name of the group
+     */
+    function createGroup(string _groupName) external {
+        uint id = groupList.length + 1;
+        require(!groups[id].exists);
```
Comment (macor161) [inline contracts/Datastore.sol]: In which case `groups[id].exists` would be true?

File: contracts/Datastore.sol [hunk 2]
```diff
@@ -329,6 +341,147 @@ contract Datastore {
...
+        require(groups[_groupId].exists);
+        if(groups[_groupId].entities[_entityIndex] != 0)
+            return groups[_groupId].entities[_entityIndex];
+    }
+
+    /**
+     * @notice Get the number of entities in a group
+     * @param _groupId Id of the group to get the count from
+     */
+    function getGroupCount(uint _groupId) public view returns(uint) {
```
Comment (macor161) [inline contracts/Datastore.sol]: I'm not sure the name is appropriate. At first glance, I thought this function returned the number of groups available in the app. Maybe `getGroupEntityCount` ?

File: src/index.ts [hunk 3]
```diff
@@ -203,6 +203,135 @@ export class Datastore {
...
+    async renameGroup(groupId: number, newGroupName: string) {
+        await this._initialize()
+
+        await this._contract.renameGroup(groupId, newGroupName)
+    }
+
+    /**
+     * Returns an array of all the groups infos
+     */
+    async getGroupsInfos() {
```
Comment (macor161) [inline src/index.ts]: I think I would simply call it `getGroups()`

File: src/index.ts [hunk 4]
```diff
@@ -203,6 +203,135 @@ export class Datastore {
...
+
+        await this._contract.getGroup(groupId)
+    }
+
+    /**
+     * Returns an entity from the `entities` array 
+     * @param groupId Id of the group to get entity from
+     * @param entityIndex Index of that entity in the `entities` array
+     */
+    async getGroupEntity(groupId: number, entityIndex: number) {
```
Comment (macor161) [inline src/index.ts]: I think this function is not necessary for the end-user. We can safely remove it unless you find a good use case

File: src/index.ts [hunk 5]
```diff
@@ -203,6 +203,135 @@ export class Datastore {
...
+        await this._initialize()
+
+        await this._contract.getGroupEntity(groupId, entityIndex)
+    }
+
+    /**
+     * Returns the number of entities in the specified group
+     * @param groupId Id of the group to get count from
+     */
+    async getGroupCount(groupId: number) {
```
Comment (macor161) [inline src/index.ts]: Same thing here, don't think this function is necessary. It really is an internal implementation helper

File: contracts/Datastore.sol (hunk 1 above)
Comment (juslar) [reply contracts/Datastore.sol]: Never. Nice catch!

File: contracts/Datastore.sol (hunk 2 above)
Comment (juslar) [reply contracts/Datastore.sol]: Totally agree. Will change.

**Comment by macor161** [comment]
Perfect! :)

**Pull request merged**